### PR TITLE
Stats: Add Exemplar class to DistributionData.

### PR DESCRIPTION
**Pull request by songy23**
[Proto definition](https://github.com/census-instrumentation/opencensus-proto/blob/master/opencensus/proto/stats/metrics/metrics.proto#L236-L250):

```
  // Exemplars are example points that may be used to annotate aggregated
  // Distribution values. They are metadata that gives information about a
  // particular value added to a Distribution bucket.
  message Exemplar {
    // Value of the exemplar point. It determines which bucket the exemplar
    // belongs to.
    double value = 1;

    // The observation (sampling) time of the above value.
    google.protobuf.Timestamp timestamp = 2;

    // Contextual information about the example value.
    map<string, string> attachments = 3;
  }
```

Next step: record and store the `Exemplar`s in the impl (in another PR).

### CODE REVIEW

File: api/src/main/java/io/opencensus/stats/AggregationData.java [hunk 1]
```diff
@@ -288,8 +293,54 @@ public static DistributionData create(
...
+      Utils.checkNotNull(exemplars, "exemplar list should not be null.");
+      Utils.checkArgument(
+          exemplars.size() <= bucketCounts.size(),
+          "each histogram bucket can have up to one exemplar, "
+              + "so the length of exemplar list should be no greater than that of bucket list.");
+      for (Exemplar exemplar : exemplars) {
+        Utils.checkNotNull(exemplar, "exemplar should not be null.");
+      }
+
+      DistributionData data =
```
Comment (dinooliva) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: Why isn't this just a return?

Comment (songy23) [reply api/src/main/java/io/opencensus/stats/AggregationData.java]: Good catch, fixed.

File: api/src/main/java/io/opencensus/stats/AggregationData.java [hunk 2]
```diff
@@ -288,8 +293,52 @@ public static DistributionData create(
         Utils.checkNotNull(bucket, "bucket should not be null.");
       }
 
+      Utils.checkNotNull(exemplars, "exemplar list should not be null.");
+      Utils.checkArgument(
+          exemplars.size() <= bucketCounts.size(),
+          "each histogram bucket can have up to one exemplar, "
+              + "so the length of exemplar list should be no greater than that of bucket list.");
```
Comment (sebright) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: I think we should either allow any number of exemplars or perform this check by comparing the exemplar values with the bucket boundaries.  The current check isn't complete, because it would allow two exemplars in one bucket if there were zero exemplars in another bucket.

File: api/src/main/java/io/opencensus/stats/AggregationData.java [hunk 3]
```diff
@@ -288,8 +293,52 @@ public static DistributionData create(
...
+
       return new AutoValue_AggregationData_DistributionData(
-          mean, count, min, max, sumOfSquaredDeviations, bucketCountsCopy);
+          mean,
+          count,
+          min,
+          max,
+          sumOfSquaredDeviations,
+          bucketCountsCopy,
+          Collections.<Exemplar>unmodifiableList(exemplars));
```
Comment (sebright) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: The exemplar list also needs to be copied.

File: api/src/main/java/io/opencensus/stats/AggregationData.java [hunk 4]
```diff
@@ -352,6 +409,61 @@ public static DistributionData create(
...
+       *
+       * @return value of the {@code Exemplar} point.
+       * @since 0.16
+       */
+      public abstract double getValue();
+
+      /**
+       * Returns the observation (sampling) time of the above value.
+       *
+       * @return the observation (sampling) time of the above value.
```
Comment (sebright) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: The meaning of "sampling" isn't very clear in this Javadoc (and `Exemplar.create`).

Since AggregationData isn't part of the metrics API, I think these exemplars should only be recorded by OpenCensus, so we could just say something like "the time that this Exemplar's value was recorded".

File: api/src/main/java/io/opencensus/stats/AggregationData.java [hunk 5]
```diff
@@ -352,6 +409,61 @@ public static DistributionData create(
...
+       *
+       * @param value value of the {@link Exemplar} point.
+       * @param timestamp the observation (sampling) time of the above value.
+       * @param attachments the contextual information about the example value.
+       * @return an {@code Exemplar}.
+       * @since 0.16
+       */
+      public static Exemplar create(
+          double value, Timestamp timestamp, Map<String, String> attachments) {
+        Utils.checkNotNull(attachments, "attachments");
```
Comment (sebright) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: AutoValue should already add null checks all of the parameters that aren't nullable, though we should check that the keys and values are non-null after making the copy.

File: api/src/test/java/io/opencensus/stats/AggregationDataTest.java [hunk 6]
```diff
@@ -57,6 +64,31 @@ public void testCreateDistributionData() {
...
   }
 
+  @Test
+  public void testCreateDistributionDataWithExemplar() {
+    Exemplar exemplar1 = Exemplar.create(4, TIMESTAMP_2, ATTACHMENTS);
+    Exemplar exemplar2 = Exemplar.create(1, TIMESTAMP_1, ATTACHMENTS);
+    DistributionData distributionData =
+        DistributionData.create(
+            7.7, 10, 1.1, 9.9, 32.2, Arrays.asList(4L, 1L), Arrays.asList(exemplar1, exemplar2));
+    assertThat(distributionData.getExemplars()).containsExactly(exemplar1, exemplar2).inOrder();
```
Comment (sebright) [inline api/src/test/java/io/opencensus/stats/AggregationDataTest.java]: Should we sort the exemplars by value?

File: api/src/test/java/io/opencensus/stats/AggregationDataTest.java [hunk 7]
```diff
@@ -57,6 +64,31 @@ public void testCreateDistributionData() {
...
+  @Test
+  public void testExemplar() {
+    Exemplar exemplar = Exemplar.create(15.0, TIMESTAMP_1, ATTACHMENTS);
+    assertThat(exemplar.getValue()).isEqualTo(15.0);
+    assertThat(exemplar.getTimestamp()).isEqualTo(TIMESTAMP_1);
+    assertThat(exemplar.getAttachments()).isEqualTo(ATTACHMENTS);
+  }
+
+  @Test
+  public void testExemplar_PreventNullAttachments() {
```
Comment (sebright) [inline api/src/test/java/io/opencensus/stats/AggregationDataTest.java]: This test suite could also test a null timestamp.

File: api/src/main/java/io/opencensus/stats/AggregationData.java [hunk 8]
```diff
@@ -352,6 +409,61 @@ public static DistributionData create(
         Function<? super AggregationData, T> defaultFunction) {
       return p3.apply(this);
     }
+
+    /**
+     * Example points that may be used to annotate aggregated distribution values, associated with a
```
Comment (sebright) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: I think point

Comment (sebright) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: I think "points" should be singular.

File: api/src/main/java/io/opencensus/stats/AggregationData.java (hunk 4 above)
Comment (sebright) [reply api/src/main/java/io/opencensus/stats/AggregationData.java]: The Javadoc for `Exemplar.create` still says "sampling".

File: api/src/main/java/io/opencensus/stats/AggregationData.java [hunk 9]
```diff
@@ -352,6 +406,65 @@ public static DistributionData create(
...
+      /**
+       * Creates an {@link Exemplar}.
+       *
+       * @param value value of the {@link Exemplar} point.
+       * @param timestamp the time that this {@code Exemplar}'s value was recorded.
+       * @param attachments the contextual information about the example value.
+       * @return an {@code Exemplar}.
+       * @since 0.16
+       */
+      public static Exemplar create(
```
Comment (dinooliva) [inline api/src/main/java/io/opencensus/stats/AggregationData.java]: Should this be public or package private? Either is fine, mostly curious about expected use.

Comment (songy23) [reply api/src/main/java/io/opencensus/stats/AggregationData.java]: I would prefer to make the factory method public. This makes it easier for any library that uses `Exemplar` (e.g exporters) to write unit tests.
